### PR TITLE
dev-cpp/notcurses: initial ebuild 2.2.3

### PR DESCRIPTION
**Pull request by dankamongmen**
This corresponds to https://bugs.gentoo.org/769875

New ebuild for Notcurses 2.2.1 in dev-cpp.

```
Hey there! I'm the upstream maintainer of Notcurses (https://notcurses.com), an ambitious library for blingful TUIs and character graphics. It's used by a few tools, and hopefully more all the time. This ebuild packages the current (2.2.1) release.

This package includes shared libraries, static libraries, pkg-config support, CMake support, header files, a few binaries built as part of Notcurses, and the data files used by those binaries. Total installation space ought be less than 2MB. The following binaries are included:

* notcurses-demo
* notcurses-input
* nctetris
* ncplayer
* ncls

"notcurses-tester" is not being built due to an absence of @onqtam's doctest in Gentoo (I'll likely package this latter up once this one gets reviewed/approved). In addition, the Python and Rust wrappers are not being built or installed (yet), nor are their associated binaries.

This is my first ebuild; please be gentle =].

hack on! --nick
```

More info:

https://notcurses.com
https://nick-black.com/dankwiki/index.php/Notcurses
https://www.youtube.com/watch?v=cYhZ7myXyyg

### CODE REVIEW

File: dev-cpp/notcurses/notcurses-2.2.1.ebuild [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+DESCRIPTION="Blingful TUIs and character graphics"
+HOMEPAGE="https://notcurses.com"
+SRC_URI="https://github.com/dankamongmen/notcurses/archive/v2.2.1.tar.gz"
+
+LICENSE="APACHE-2.0"
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+
+DEPEND="app-text/pandoc dev-libs/unistring media-video/ffmpeg sys-libs/readline"
```
Comment (Polynomial-C) [inline dev-cpp/notcurses/notcurses-2.2.1.ebuild:17]: One package per line please

File: dev-cpp/notcurses/notcurses-2.2.1.ebuild [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
+
+CMAKE_ECLASS="cmake"
+inherit cmake-multilib
+
+DESCRIPTION="Blingful TUIs and character graphics"
+HOMEPAGE="https://notcurses.com"
+SRC_URI="https://github.com/dankamongmen/notcurses/archive/v2.2.1.tar.gz"
```
Comment (Polynomial-C) [inline dev-cpp/notcurses/notcurses-2.2.1.ebuild:11]: Please replace this with
```
SRC_URI="https://github.com/dankamongmen/${PN}/archive/v${PV}.tar.gz -> ${P}.tar.gz"
```

File: dev-cpp/notcurses/notcurses-2.2.1.ebuild [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+src_prepare() {
+	cmake_src_prepare
+}
+
+src_configure() {
+	local mycmakeargs=(
+		-DUSE_DOCTEST=OFF
+		-DUSE_QRCODEGEN=OFF
+		-DUSE_PANDOC=OFF
```
Comment (Polynomial-C) [inline dev-cpp/notcurses/notcurses-2.2.1.ebuild:29]: You have `app-text/pandoc` in `DEPEND` but yet disable pandoc usage. Why?

**Review by Polynomial-C — CHANGES_REQUESTED** [review]
You may consider using the `doc` use-flag to trigger pandoc usage. pandoc is a huge dependency which most Gentoo users would not want to install an entire haskell stack for.

File: dev-cpp/notcurses/notcurses-2.2.1.ebuild [hunk 4]
```diff
@@ -0,0 +1,40 @@
+# Copyright 2021 nick black
```
Comment (hydrapolic) [inline dev-cpp/notcurses/notcurses-2.2.1.ebuild:1]: Gentoo Authors ?

Comment (dankamongmen) [reply dev-cpp/notcurses/notcurses-2.2.1.ebuild]: Done, thanks.

File: dev-cpp/notcurses/notcurses-2.2.1.ebuild (hunk 1 above)
Comment (dankamongmen) [reply dev-cpp/notcurses/notcurses-2.2.1.ebuild]: Done, thanks.

File: dev-cpp/notcurses/notcurses-2.2.1.ebuild (hunk 3 above)
Comment (dankamongmen) [reply dev-cpp/notcurses/notcurses-2.2.1.ebuild:32]: Brain freeze. I'll make this conditional based on the user selecting `doc`, as you suggested.

File: dev-cpp/notcurses/notcurses-2.2.1.ebuild (hunk 2 above)
Comment (dankamongmen) [reply dev-cpp/notcurses/notcurses-2.2.1.ebuild]: Done, thanks. FWIW, the packaging guide recommends against this usage (https://wiki.gentoo.org/wiki/Basic_guide_to_write_Gentoo_Ebuilds).

**Comment by dankamongmen** [comment]
There are several issues being reported by CI, which I intend to address today.

**Comment by dankamongmen** [comment]
> You may consider using the `doc` use-flag to trigger pandoc usage. pandoc is a huge dependency which most Gentoo users would not want to install an entire haskell stack for.

Done.

**Comment by dankamongmen** [comment]
> There are several issues being reported by CI, which I intend to address today.

Got the "unresolveable dep" fixed for `dev-libs/unistring` (should have been `libunistring`)

**Comment by dankamongmen** [comment]
* Changed `docs` to `doc` (unknown USE flag: 'docs')
* Renaming to notcurses-${P}.tar.gz (bad filename: [ v2.2.1.tar.gz ])

**Comment by dankamongmen** [comment]
Alright, I'm down to two warnings:

* "UnnecessaryManifest	unnecessary file in Manifest: [ notcurses-2.2.1.ebuild ]"
* "PkgMissingMetadataXml	package is missing metadata.xml"

Googling around seems to suggest that I need switch to "thin manifests". I'm going to add the `metadata.xml` file now.

**Comment by dankamongmen** [comment]
...and that last one ought convert to thin manifest, and eliminate the last warning.

**Comment by dankamongmen** [comment]
Adding `metadata.xml` earned a new warning. I've added the proxy-maintainers entry, which will hopefully resolve this latest problem. =]

**Comment by dankamongmen** [comment]
Alright, all checks are now passing, PTAL @hydrapolic or @Polynomial-C  (or anyone else). Thanks for your patience!

**Comment by dankamongmen** [comment]
no-signoff needs to come off, and bug-linked can be added

File: dev-cpp/notcurses/metadata.xml [hunk 5]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE pkgmetadata SYSTEM "http://www.gentoo.org/dtd/metadata.dtd">
+<pkgmetadata>
+<maintainer type="person">
+<name>nick black</name>
```
Comment (hydrapolic) [inline dev-cpp/notcurses/metadata.xml:5]: Please indent for better readability

File: dev-cpp/notcurses/notcurses-2.2.1.ebuild [hunk 6]
```diff
@@ -0,0 +1,43 @@
+# Copyright 2021 Gentoo Authors
```
Comment (hydrapolic) [inline dev-cpp/notcurses/notcurses-2.2.1.ebuild:1]: # Copyright 1999-2021 Gentoo Authors

File: dev-cpp/notcurses/metadata.xml (hunk 5 above)
Comment (dankamongmen) [reply dev-cpp/notcurses/metadata.xml:5]: Done.

File: dev-cpp/notcurses/notcurses-2.2.1.ebuild (hunk 6 above)
Comment (dankamongmen) [reply dev-cpp/notcurses/notcurses-2.2.1.ebuild:1]: Done.

**Comment by dankamongmen** [comment]
Rebased against tip of master

**Review by Polynomial-C — APPROVED** [review]
Looking good now, thanks.

**Comment by dankamongmen** [comment]
I went ahead and rebased, plus updated this to 2.2.2. I'd love to see it merged! =]

**Comment by dankamongmen** [comment]
hey @Polynomial-C , you approved this 20 days ago -- is there anything blocking it from being merged? sorry to be annoying, just hoping to get some other stuff packaged for gentoo that depends on this.

File: dev-cpp/notcurses/notcurses-2.2.2.ebuild [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+LICENSE="Apache-2.0"
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+IUSE="doc"
+
+DEPEND="dev-libs/libunistring
+	media-video/ffmpeg
+	>=sys-libs/readline-8.0"
+RDEPEND="${DEPEND}"
+BDEPEND="doc? ( app-text/pandoc )"
```
Comment (juippis) [inline dev-cpp/notcurses/notcurses-2.2.2.ebuild:22]: As you are the upstream, you'd do a huge favor for everyone by shipping pre-generated doc files, if possible. I see `find_package(PkgConfig REQUIRED)` so `BDEPEND` on `virtual/pkgconfig` too.

File: dev-cpp/notcurses/notcurses-2.2.2.ebuild [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+	cmake_src_prepare
+}
+
+src_configure() {
+	local mycmakeargs=(
+		-DUSE_DOCTEST=OFF
+		-DUSE_QRCODEGEN=OFF
+		-DUSE_PANDOC="$(usex doc)"
+	)
+	cmake-multilib_src_configure
```
Comment (juippis) [inline dev-cpp/notcurses/notcurses-2.2.2.ebuild:34]:
```
if(${USE_FFMPEG})
pkg_check_modules(AVCODEC REQUIRED libavcodec>=57.0)
pkg_check_modules(AVFORMAT REQUIRED libavformat>=57.0)
pkg_check_modules(AVUTIL REQUIRED libavutil>=56.0)
pkg_check_modules(SWSCALE REQUIRED libswscale>=5.0)
set_property(GLOBAL APPEND PROPERTY PACKAGES_FOUND FFMpeg)
elseif(${USE_OIIO})
pkg_check_modules(OIIO REQUIRED OpenImageIO>=2.1)
set_property(GLOBAL APPEND PROPERTY PACKAGES_FOUND OpenImageIO)
```
You could make it visible here that ffmpeg is the chosen configure flag, but that openimageio is also an option. Gentoo users do love the choice ;) but imho you are doing correctly just choosing ffmpeg without any ugly `REQUIRED_USE` restrictions.

**Review by juippis — COMMENTED** [review]
I installed this with `USE="doc"`, but this is what I get:
```
~ # equery f notcurses | grep doc
/usr/share/doc
/usr/share/doc/notcurses-2.2.2
/usr/share/doc/notcurses-2.2.2/NEWS.md.bz2
/usr/share/doc/notcurses-2.2.2/README.md.bz2
/usr/share/doc/notcurses-2.2.2/TERMINALS.md.bz2
/usr/share/doc/notcurses-2.2.2/USAGE.md.bz2
```
Maybe you need to manually install them in `multilib_src_install_all()`?

Then please, format your commit message as:
```dev-cpp/notcurses: initial ebuild 2.2.2

notcurses is <explain shortly what this package is about>

Signed-off-by: 
```

**Comment by juippis** [comment]
Also by default disable building static-libraries, or remove the static library files in `multilib_src_install_all()`.

File: dev-cpp/notcurses/notcurses-2.2.2.ebuild (hunk 7 above)
Comment (dankamongmen) [reply dev-cpp/notcurses/notcurses-2.2.2.ebuild:22]: I generate documentation packs with each release, e.g.

https://github.com/dankamongmen/notcurses/releases/download/v2.2.3/notcurses-doc-2.2.3.tar.gz

could you suggest how this tarball can be best integrated with the package overall? Thanks!

Comment (thesamesam) [reply dev-cpp/notcurses/notcurses-2.2.2.ebuild:22]: We can add it into SRC_URI! It accepts multiple (space-separated) elements.

**Comment by dankamongmen** [comment]
> I installed this with `USE="doc"`, but this is what I get:
> EDIT: I do see a lot of files under `/usr/share/man` though.

yeah, if you got the several dozen man pages, you built with `doc` enabled, and got everything it intended.

> Then please, format your commit message as:
> 
> ```
> dev-cpp/notcurses: initial ebuild 2.2.2
> 
> notcurses is <explain shortly what this package is about>
> 
> Signed-off-by: 
> ```

done. i'm updating the ebuild to 2.2.3, but i've added the description.

File: dev-cpp/notcurses/notcurses-2.2.2.ebuild [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
+
+CMAKE_ECLASS="cmake"
+inherit cmake-multilib
+
+DESCRIPTION="Blingful TUIs and character graphics"
+HOMEPAGE="https://notcurses.com"
+SRC_URI="https://github.com/dankamongmen/${PN}/archive/v${PV}.tar.gz -> ${P}.tar.gz"
```
[the remark this answers is not in the document]
Comment (dankamongmen) [reply dev-cpp/notcurses/notcurses-2.2.2.ebuild]: heh, i read as much on the Gentoo documentation, but previous reviewers asked that i use this (discouraged) method. i've pulled it. Done.

File: dev-cpp/notcurses/notcurses-2.2.2.ebuild [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+DEPEND="dev-libs/libunistring
+	media-video/ffmpeg
+	>=sys-libs/readline-8.0"
+RDEPEND="${DEPEND}"
+BDEPEND="doc? ( app-text/pandoc )"
+
+src_prepare() {
+	cmake_src_prepare
+}
```
[the remark this answers is not in the document]
Comment (dankamongmen) [reply dev-cpp/notcurses/notcurses-2.2.2.ebuild]: Done, verified that it still builds, thanks.

File: dev-cpp/notcurses/notcurses-2.2.2.ebuild (hunk 7 above)
Comment (dankamongmen) [reply dev-cpp/notcurses/notcurses-2.2.2.ebuild]: cool, i've done that, but then...where do i want to "manually" install these from?

Comment (dankamongmen) [reply dev-cpp/notcurses/notcurses-2.2.2.ebuild]: also, i see i can make `SOURCE_URI` components dependent upon use variables. ought i make these dependent on `doc`? i.e. is `doc` about the cost of building documentation, or the cost of installing it?

**Comment by dankamongmen** [comment]
> Also by default disable building static-libraries, or remove the static library files in `multilib_src_install_all()`.

no problem; i'll add `USE_STATIC=off`.

**Comment by dankamongmen** [comment]
> no problem; i'll add `USE_STATIC=off`.

Done.

File: dev-cpp/notcurses/notcurses-2.2.2.ebuild (hunk 8 above)
Comment (dankamongmen) [reply dev-cpp/notcurses/notcurses-2.2.2.ebuild]: Sorry, but I'm not sure if/what you're asking me to do here :/.

File: dev-cpp/notcurses/notcurses-2.2.2.ebuild (hunk 7 above)
Comment (juippis) [reply dev-cpp/notcurses/notcurses-2.2.2.ebuild]: I'd install them unconditionally. It's 1 MB file. 

Here it's about the dependencies required to *build* those 1 MB worth of files, installing *pandoc* on my system requires 135 new packages to be installed.

File: dev-cpp/notcurses/notcurses-2.2.2.ebuild (hunk 8 above)
Comment (juippis) [reply dev-cpp/notcurses/notcurses-2.2.2.ebuild]: I basically wanted you to make the configure options visible, similarily to how 
```
	-DUSE_DOCTEST=OFF
		-DUSE_QRCODEGEN=OFF
		-DUSE_STATIC=OFF
```
is currently visible in your ebuild. So people know about these configure options. We have `EXTRA_ECONF` available to control "hard-coded" variables for more user choice.

File: dev-cpp/notcurses/notcurses-2.2.3.ebuild [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+CMAKE_ECLASS="cmake"
+inherit cmake-multilib
+
+DESCRIPTION="Blingful TUIs and character graphics"
+HOMEPAGE="https://notcurses.com"
+SRC_URI="https://github.com/dankamongmen/notcurses/archive/v${PV}.tar.gz -> ${P}.tar.gz https://github.com/dankamongmen/notcurses/releases/download/v${PV}/notcurses-doc-${PV}.tar.gz"
+LICENSE="Apache-2.0"
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+IUSE="doc"
```
Comment (juippis) [inline dev-cpp/notcurses/notcurses-2.2.3.ebuild:15]: Without me trying this new ebuild, what does `doc` do to the ebuild now? How do you install the downloaded extra documentation?

File: dev-cpp/notcurses/notcurses-2.2.3.ebuild [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+DEPEND="dev-libs/libunistring
+	media-video/ffmpeg
+	>=sys-libs/readline-8.0"
+RDEPEND="${DEPEND}"
+BDEPEND="doc? ( app-text/pandoc )"
```
Comment (juippis) [inline dev-cpp/notcurses/notcurses-2.2.3.ebuild:21]: In other words, you already have the pre-generated documentation available, thus we don't need to depend on *pandoc* anymore right?

**Review by juippis — COMMENTED** [review]
We're nearly there! Thanks for patience and cooperation.

I peeked inside **notcurse-doc-2.3.3.tar.gz** and you have `.so` files in it. Is that intentional? You'll most likely have to work a bit in the ebuild to get those to be installed in correct directories. Ie, manpages and .html pages to separate places.

You do have a Gentoo system to test these right? You can use *equery -f* from *app-portage/gentoolkit* to list installed files of a certain package.

File: dev-cpp/notcurses/notcurses-2.2.2.ebuild (hunk 7 above)
Comment (dankamongmen) [reply dev-cpp/notcurses/notcurses-2.2.2.ebuild]: Got it. We're unconditionally downloading the file. I'll go ahead and extract it in ... i guess i unpack it in the prepare step, and install in the install step?

File: dev-cpp/notcurses/notcurses-2.2.3.ebuild (hunk 12 above)
Comment (dankamongmen) [reply dev-cpp/notcurses/notcurses-2.2.3.ebuild:21]: yep, i'll make that change, thanks.

File: dev-cpp/notcurses/notcurses-2.2.3.ebuild (hunk 11 above)
Comment (dankamongmen) [reply dev-cpp/notcurses/notcurses-2.2.3.ebuild:15]: as i'm envisioning things, `doc` would cover installation of the man pages from the doc package (but not download of said package). that's what i intend to implement. i'm also happy to unconditionally install the man pages.

**Comment by dankamongmen** [comment]
> I peeked inside **notcurse-doc-2.3.3.tar.gz** and you have `.so` files in it. Is that intentional? You'll most likely have to work a bit in the ebuild to get those to be installed in correct directories. Ie, manpages and .html pages to separate places.

it most certainly is not. i'll likely repack the archive, unless i cut 2.3.4 in the very near future. thanks for the heads-up!

> You do have a Gentoo system to test these right? You can use _equery -f_ from _app-portage/gentoolkit_ to list installed files of a certain package.

yep, i do, i just didn't inspect the unpacked doc package. i don't think i've ever used it, heh.

**Comment by dankamongmen** [comment]
> > I peeked inside **notcurse-doc-2.3.3.tar.gz** and you have `.so` files in it. Is that intentional? You'll most likely have to work a bit in the ebuild to get those to be installed in correct directories. Ie, manpages and .html pages to separate places.
> 
> it most certainly is not. i'll likely repack the archive, unless i cut 2.3.4 in the very near future. thanks for the heads-up!

I've updated the 2.2.3 doc tarball manually, and fixed the issue in my release process. Thanks again!

File: dev-cpp/notcurses/notcurses-2.2.3.ebuild (hunk 12 above)
Comment (dankamongmen) [reply dev-cpp/notcurses/notcurses-2.2.3.ebuild]: I've removed the `BDEPEND` and also the conditional `-DUSE_PANDOC=on`. Done.

File: dev-cpp/notcurses/notcurses-2.2.3.ebuild (hunk 11 above)
Comment (dankamongmen) [reply dev-cpp/notcurses/notcurses-2.2.3.ebuild:15]: by the way, you mention installing .xhtml files somewhere, but i'm not intended to do that. the only files i'd copy out of the doc tarball would be the man pages. the .xhtml files have a bijection on the man pages, and aren't really useful IMHO on a development workstation (they're intended more for a web server).

Comment (dankamongmen) [reply dev-cpp/notcurses/notcurses-2.2.3.ebuild:15]: so what i'm planning to do is: add a `src_install` function, which will run `cmake-multilib_src_install`, and use the `doman` command to install the man pages, iff the `doc` variable is set. does that sound about right?

File: dev-cpp/notcurses/notcurses-2.2.2.ebuild (hunk 8 above)
Comment (dankamongmen) [reply dev-cpp/notcurses/notcurses-2.2.2.ebuild]: Done.

File: dev-cpp/notcurses/notcurses-2.2.2.ebuild (hunk 7 above)
Comment (dankamongmen) [reply dev-cpp/notcurses/notcurses-2.2.2.ebuild]: I have made download of the doc package conditional on doc using the syntax from https://devmanual.gentoo.org/ebuild-writing/variables/. I'm going to conditionally install them from the `src_install` step using the `doman` function. I need to test this with `doc` set, but I think we're good. Thanks for the feedback!

File: dev-cpp/notcurses/notcurses-2.2.3.ebuild (hunk 11 above)
Comment (dankamongmen) [reply dev-cpp/notcurses/notcurses-2.2.3.ebuild]: I've got this all written out now, and just need to test with `doc` in place. I'm not certain that I'm unpacking the doc tarball to a place where `doman` is going to find it. Everything seems to work in the no-`doc` case.

**Comment by dankamongmen** [comment]
OK, I think I've dealt with all comments. I'm now conditionally installing the man pages as unpacked from the doc tarball when `doc` is set, but I don't see them following `ebuild install`. Here's my `src_instal()` function:

```sh
src_install() {
	cmake-multilib_src_install
	if use doc ; then
		# we use this tortured form lest we try, every time we release a
		# x.y.1 or x.y.3, to install the source dir as a man page.
		# exploit the fact that there's a bijection from html<>man.
		for i in ../*.html ; do
			doman ../$(basename $i .html)
		done
	fi
}
```

when i put a `ls ../$(basename $i .html)` above the `doman`, i see the source man page exactly as i expect, and `doman` is not failing, but i'm not seeing them installed on to my system...

once this is done, i think we're ready to go =].

**Review by juippis — APPROVED** [review]
Thanks, all seems fine! I'm going to modify the ebuild a bit so it always installs the manpages. The tarball is only 88 kb, I can't justify a USE flag to control installing such important core docs for ~150 kb worth.

**Comment by dankamongmen** [comment]
> Thanks, all seems fine! I'm going to modify the ebuild a bit so it always installs the manpages. The tarball is only 88 kb, I can't justify a USE flag to control installing such important core docs for ~150 kb worth.

Please feel free; I only made it conditional because I thought that to be what y'all wanted =]. Thanks so much for patiently walking me through the new ebuild process -- godspeed, new little ebuild!

**Pull request closed without merging**